x-pack/plugin/snapshot-repo-test-kit/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalysisSuccessIT.java

```diff
@@ -442,10 +442,13 @@ public Map<String, BlobMetadata> listBlobsByPrefix(OperationPurpose purpose, Str
         @Override
         public void getRegister(OperationPurpose purpose, String key, ActionListener<OptionalBytesReference> listener) {
             assertPurpose(purpose);
-            if (firstRegisterRead.compareAndSet(true, false) && randomBoolean() && randomBoolean()) {
+            if (key.startsWith(RepositoryAnalyzeAction.CONTENDED_REGISTER_NAME_PREFIX)
+                && firstRegisterRead.compareAndSet(true, false)
+                && randomBoolean()
+                && randomBoolean()) {
                 // only fail the first read, we must not fail the final check
                 listener.onResponse(OptionalBytesReference.EMPTY);
-            } else if (randomBoolean()) {
+            } else if (key.startsWith(RepositoryAnalyzeAction.UNCONTENDED_REGISTER_NAME_PREFIX) || randomBoolean()) {
                 listener.onResponse(OptionalBytesReference.of(registers.computeIfAbsent(key, ignored -> new BytesRegister()).get()));
             } else {
                 final var bogus = randomFrom(BytesArray.EMPTY, new BytesArray(new byte[] { randomByte() }));
```

Comment:
This logic here is a bit to follow. IIUC, we don't want to return anything wrong for uncontended analysis. But the code here seems to suggest that we could be returning bogus result for it. But not really because the call to `compareAndExchangeRegister` always return a `new BytesRegister()` which always return 0 regardless of the bogus value. I think it would be better if we could make this more explicit for the uncontended operations. Maybe have it as the top level switch, e.g.:
```
if (key.startsWith(RepositoryAnalyzeAction.UNCONTENDED_REGISTER_NAME_PREFIX)) {
    ...
} else {
    // everything else, essentially the existing code
}
```
?

Reply:
++ yes this is all a little unsatisfactory indeed. This area is going to need a little rework once #101184 is merged, I'll try and bring the checks on the key prefix to the top level in these methods.

Reply:
Ok now merged and cleaned up, see 2dd4250.